[QoS] Add support for handling empty `EVM` responses

adshmh opened the pull request:
## Summary
Add empty response handling and validation reason tracking for EVM QoS

### Primary Changes:
- Added `EVMResponseInvalidReason` enum to track specific reasons for endpoint response validation failures (e.g. `EMPTY_RESPONSE`, `UNMARSHAL_ERR`)
- Implemented empty response handling with automatic endpoint disqualification through new `EVMEmptyResponse` type
- Added `invalid_response_reason` label to Prometheus metrics for monitoring QoS validation failures
- Updated protobuf definitions and generated Go code

### Secondary changes:
- Added `getEVMEndpointResponseValidationFailureReason` helper function
- Added TODO notes for future retry functionality implementation

## Issue

- Description: Returning an empty response lowers QoS for users, and any endpoint returning such a response should be dropped from the selection pool.

- Issue: #{ISSUE_NUMBER}

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Sanity Checklist

- [x] I have updated the GitHub Issue `assignees`, `reviewers`, `labels`, `project`, `iteration` and `milestone`
- [ ] For docs, I have run `make docusaurus_start`
- [ ] For code, I have run `make test_all`
- [ ] For configurations, I have update the documentation
- [x] I added TODOs where applicable

Olshansk commented on metrics/qos/evm.go:
```diff
@@ -78,3 +85,30 @@ func getEVMRequestSuccess(evmObservations *qos.EVMRequestObservations) bool {
 	return false
 }
+
+// IN_THIS_PR: Add Grafana panel(s) to visualize the validation reason label in Local development mode.
```
```suggestion
// TODO_IN_THIS_PR: Add Grafana panel(s) to visualize the validation reason label in Local development mode.
```

Olshansk commented on metrics/qos/evm.go:
```diff
@@ -78,3 +85,30 @@ func getEVMRequestSuccess(evmObservations *qos.EVMRequestObservations) bool {
+// aggregated observations.
+//
+// getEVMEndpointResponseValidationFailureReason returns why the endpoint response failed QoS validation.
+func getEVMEndpointResponseValidationFailureReason(evmObservations *qos.EVMRequestObservations) string {
```
```suggestion
func getEVMEndpointResponseValidationFailureReason(
	evmObservations *qos.EVMRequestObservations,
) string {
```

Olshansk commented on proto/path/qos/evm.proto:
```diff
@@ -6,6 +6,16 @@ option go_package = "github.com/buildwithgrove/path/observation/qos";
+// Current invalid response types (as of PR #164):
+//   1. EmptyResponse - endpoint returned no data
+//   2. UnmarshalErr - response failed to parse into expected format
+enum EVMResponseInvalidReason {
+ REASON_UNSPECIFIED = 0;
+ REASON_EMPTY_RESPONSE = 1; // No response data received
+ REASON_UNMARSHAL_ERR = 2;  // Response parsing failed
+}
```
```suggestion
enum EVMResponseInvalidReason {
	EVM_RESPONSE_INVALID_REASON_UNSPECIFIED = 0;
	EVM_RESPONSE_INVALID_REASON_EMPTY = 1;      // No response data received
	EVM_RESPONSE_INVALID_REASON_UNMARSHAL = 2;  // Response parsing failed
}
```

The main reason I'm suggesting this:

<img width="653" alt="Screenshot 2025-02-13 at 9 10 58 PM" src="https://github.com/user-attachments/assets/0eda7d14-c314-4172-9e15-156f75e3311c" />

Olshansk commented on qos/evm/response_empty.go:
```diff
@@ -0,0 +1,55 @@
+// emptyResponse provides the functionality required from a response by a requestContext instance.
+var _ response = responseEmpty{}
+
+// TODO_MVP(@adshmh): Implement request retry support:
```
This is really great!

Olshansk commented on qos/evm/response_empty.go:
```diff
@@ -0,0 +1,55 @@
+//  3. Extend evm.response interface with ShouldRetry()
+//  4. Add ShouldRetry() to evm.requestContext to evaluate retry eligibility based on responses
+//
+// responseEmpty processes empty endpoint responses by:
```
Love how easy this is to read!

adshmh replied:
Credit to Claude for rewriting most of the comments!

Olshansk commented on qos/evm/response_blocknumber.go:
```diff
@@ -92,7 +106,7 @@ func (r responseToBlockNumber) GetResponsePayload() []byte {
 	if err != nil {
 		// This should never happen: log an entry but return the response anyway.
-		r.logger.Warn().Err(err).Msg("responseToGetHealth: Marshaling JSON-RPC response failed.")
+		r.logger.Warn().Err(err).Msg("responseToBlockNumber: Marshaling JSONRPC response failed.")
```
nice

Olshansk commented on qos/evm/preset_responses.go:
```diff
@@ -0,0 +1,27 @@
+	"github.com/buildwithgrove/path/qos/jsonrpc"
+)
+
+// errResponseEmptyEndpointResponse represents a JSON-RPC error response for when
+// an endpoint returns an empty response body.
+// This indicates server misbehavior, triggers removal of the endpoint from the selection pool,
+// and tells the client they can retry with a different endpoint.
```
```suggestion
// errResponseEmptyEndpointResponse is a pre-defined JSON-RPC response structure to be used (i.e. returned)
// for all EVM JSON-RPC requests when the endpoint returned an empty response body.
// This indicates server misbehavior, triggers removal of the endpoint from the selection pool,
// and tells the client they can retry with a different endpoint.
```

Olshansk commented on qos/evm/preset_responses.go:
```diff
@@ -0,0 +1,27 @@
+// and tells the client they can retry with a different endpoint.
+var errResponseEmptyEndpointResponse = jsonrpc.GetErrorResponse(
+	jsonrpc.ID{}, // Use request's original ID if present
+	-32000,       // JSON-RPC server error code
```
```suggestion
	-32000,       // JSON-RPC standard server error code; https://www.jsonrpc.org/historical/json-rpc-2-0.html
```

Olshansk commented on qos/evm/preset_responses.go:
```diff
@@ -0,0 +1,27 @@
+var errResponseEmptyEndpointResponse = jsonrpc.GetErrorResponse(
+	jsonrpc.ID{}, // Use request's original ID if present
+	-32000,       // JSON-RPC server error code
+	"Server error: Received an empty response. Server will be dropped from the selection pool. Please try again.", // Error Response Message
```
Wdyt of this instead? Server has a few too many meanings here.

```suggestion
	"Endpoint (data/service node error): Received an empty response. The endpoint will be dropped from the selection pool. Please try again.", // Error Response Message
```

Olshansk commented on qos/evm/preset_responses.go:
```diff
@@ -0,0 +1,27 @@
+	-32000,       // JSON-RPC server error code
+	"Server error: Received an empty response. Server will be dropped from the selection pool. Please try again.", // Error Response Message
+	map[string]string{
+		"retryable": "true",
```
#PUC What this means

Olshansk replied:
I found `// Marks the error as retryable to allow clients to safely retry their request` below.

Let's copy-paste that here.

Olshansk reviewed (approved):
@adshmh What an amazing, useful, easy to read and well scoped PR!

1. I added a few NITs, so PTAL. Feel free to merge right after so you're not blocked on deployment.

2. As a meta point - this is beginning to remind me of the Waymo codebase with these "special cases" that are also well structured.

3. As another meta point - at some point (not now) I'd want to discuss what made QoS customization/evolution not very extensible in portal-middleware to avoid the same mistake here. For now, I believe the structure and iteration is helping uncover what we should build.

Ty for getting this done so fast!

commoddity reviewed (commented):
Thanks for the PR @adshmh .

I don't have time for a full review this morning as I'm OOO the rest of the day but I did have time to do a bit of manual testing comparing this PR to the `protocol/morse` level approach I tried as an experiment yesterday. I've posted the results in Discord, PTAL @ [the message here.] (https://discord.com/channels/824324475256438814/1273320783547990160/1339895936280956972) Hopefully this is a helpful data point for yourself and @Olshansk .

adshmh commented:
Made a few changes while fixing a bug I encountered when adding a Grafana dashboard for the new Response Validation Error:
- Added a TODO for `responseNone` type to track protocol-level failures where no endpoint response is received
- Moved EVM metrics to dedicated package `metrics/qos/evm` to improve organization and naming.
- Added response extraction helpers in `metrics/qos/evm` to simplify metric collection
- Created Grafana panel to visualize response validation errors by reason

commoddity reviewed (approved):
LGTM, have tested manually extensively as per our discussion on Discord.

Olshansk commented on proto/path/qos/evm.proto:
```diff
@@ -6,6 +6,16 @@ option go_package = "github.com/buildwithgrove/path/observation/qos";
+}
+
 // EVMRequestObservations captures all observations made while serving a single EVM blockchain service request.
 message EVMRequestObservations {
```
Note that the indent of `chain_id` here seems to be inconsistent with other parts

adshmh replied:
Thanks, ran all `.proto` files through `buf`: it only generated a warning re package naming.
Added a TODO to address the proto files package names warnings.

Olshansk commented on qos/evm/response_empty.go:
```diff
@@ -0,0 +1,55 @@
+	return qosobservations.EVMEndpointObservation{
+		ResponseObservation: &qosobservations.EVMEndpointObservation_EmptyResponse{
+			EmptyResponse: &qosobservations.EVMEmptyResponse{
+				Valid:         false, // Empty responses are inherently invalid - explicitly set for clarity
```
Great comment

Olshansk commented on metrics/qos/evm/response.go:
```diff
@@ -0,0 +1,40 @@
+
+// response defines methods needed for response metrics collection.
+// Abstracts proto-specific details from metrics logic.
+// Note: Update when adding new metric requirements.
```
```suggestion
// DEV_NOTE: You MUST update this when adding new metric requirements.
```

Olshansk commented on metrics/qos/evm/response.go:
```diff
@@ -0,0 +1,40 @@
+
+// getResponseFromObservation extracts the response data from an endpoint observation.
+// Returns nil if no response is present.
+// Note: Update when adding new response types.
```
```suggestion
// DEV_NOTE: You MUST update this when adding new metric requirements.
```

Olshansk replied:
One day we'll change `DEV_NOTE` to `AI_AGENT_DO_THIS` ;)

Olshansk commented on proto/path/qos/evm.proto:
```diff
@@ -6,6 +6,16 @@ option go_package = "github.com/buildwithgrove/path/observation/qos";
 import "path/qos/jsonrpc.proto";
```
I'm a fan of the structure / direction this is taking.

Thinking out loud (Non-actionable): I'm curious if we end up keeping it as is or re-factoring in the future, and can actually see it going in either direciton.

Olshansk commented on metrics/qos/evm/response.go:
```diff
@@ -0,0 +1,40 @@
+		return nil
+	}
+
+	if response := observation.GetChainIdResponse(); response != nil {
```
Can you do this for every if? I know it's obvious from the code, but much easier to read.


```suggestion
	// handle chain_id response 
	if response := observation.GetChainIdResponse(); response != nil {
```

Olshansk reviewed (approved):
@adshmh Preliminary approval since there's nothing blocking, but I did leave a few NITs/edit regarding comments & naming.

1. PTAL and make the updates (or ignore/resolve) before merging.
2. I've added screenshots of how I tested this. Is this the same approach that you took or something else?

![Screenshot 2025-02-18 at 12 37 25 PM](https://github.com/user-attachments/assets/460dadd5-9fa6-44e7-b5fd-79120f923f8a)
![Screenshot 2025-02-18 at 12 37 08 PM](https://github.com/user-attachments/assets/93b2c354-434c-46c4-85bc-03def341196c)

adshmh commented on qos/evm/preset_responses.go:
```diff
@@ -0,0 +1,28 @@
+package evm
```
Renamed to `errors.go` for now, we can change that in PR #165 if needed.
Here is Claude's rationale for suggesting `errors.go` based on the content of the file:
```
Based on the content, I'd suggest naming the file:
errors.go
Rationale:

The file contains only error response constructors
All functions are prefixed with newErrResponse
Each function creates different types of JSON-RPC error responses
Follows Go convention of using simple, descriptive filenames
Matches common Go pattern of putting error-related code in errors.go
```

adshmh commented:
> 2\. I've added screenshots of how I tested this. Is this the same approach that you took or something else?

Yes, I used the same approach.
